buffer.go

```diff
@@ -380,27 +371,53 @@ func (p *bufferPool) put(b *buffer) {
 	if b.refCount() != 0 {
 		panic("BUG: buffer returned to pool with a non-zero reference count")
 	}
-	// if this slice is somehow less then our min pool size, just drop it
-	size := cap(b.data)
-	if size < basePoolIncrement {
-		return
+	if bucketIndex, _ := bufferPoolBucketIndexAndSizeOfPut(cap(b.data)); bucketIndex >= 0 {
+		p.buckets[bucketIndex].Put(b)
+	}
+}
+
+const (
+	bufferPoolBucketCount         = 32
+	bufferPoolMinSize             = 4096
+	bufferPoolLastShortBucketSize = 262144
+)
+
+func bufferPoolNextSize(size int) int {
+	if size < bufferPoolLastShortBucketSize {
+		return size * 2
+	} else {
+		return size + (size / 2)
 	}
-	i := levelledPoolIndex(size / 2) // divide by 2 to put the buffer in the level below so it will always be large enough
-	p.bucket[i].Put(b)
 }
 
-// levelledPoolIndex returns the index of the pool to use for a buffer of the
-// given size. It never returns an index that will panic.
-func levelledPoolIndex(size int) int {
-	i := size / basePoolIncrement
-	i = 32 - bits.LeadingZeros32(uint32(i)) // log2
-	if i >= numPoolBuckets {
-		i = numPoolBuckets - 1
+func bufferPoolBucketIndexAndSizeOfGet(size int) (int, int) {
+	limit := bufferPoolMinSize
+
+	for i := 0; i < bufferPoolBucketCount; i++ {
+		if size <= limit {
+			return i, limit
+		}
+		limit = bufferPoolNextSize(limit)
 	}
-	if i < 0 {
-		i = 0
+
+	return -1, size
+}
+
+func bufferPoolBucketIndexAndSizeOfPut(size int) (int, int) {
+	// When releasing buffers, some may have a capacity that is not one of the
+	// bucket sizes (due to the use of append for example). In this case, we
+	// have to put the buffer is the highest bucket with a size less or equal
+	// to the buffer capacity.
+	if limit := bufferPoolMinSize; size >= limit {
```

Comment:
I think I'd prefer to throw away the largest buffers until we have a clear measure that we need to retain them (and maybe then we'll just increase the bucket count?). Retaining large memory blobs tend to create more problems in my experience, it reduces the GC frequency because GC triggers based on the amount of memory allocated, and we can end up wasting a lot of memory if we keep buffers in the pool that are much larger than needed -- the size trends towards the largest buffers retained in general, which is why we need this bucketing strategy.